oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServerOAuth2AuthorizedClientExchangeFilterFunction.java

```diff
@@ -330,80 +420,102 @@ public void setAccessTokenExpiresSkew(Duration accessTokenExpiresSkew) {
+	 * 	       that is active for the given request.
+	 */
+	private Mono<Optional<ServerWebExchange>> effectiveServerWebExchange(ClientRequest request) {
+		return Mono.justOrEmpty(serverWebExchange(request))
+				.switchIfEmpty(currentServerWebExchangeMono)
 				.map(Optional::of)
 				.defaultIfEmpty(Optional.empty());
+	}
+
+	private Mono<OAuth2AuthorizeRequest> reauthorizeRequest(ClientRequest request, OAuth2AuthorizedClient authorizedClient) {
+		Mono<Optional<ServerWebExchange>> serverWebExchange = effectiveServerWebExchange(request);
 
-		return Mono.zip(authentication, serverWebExchange)
+		return Mono.zip(currentAuthenticationMono, serverWebExchange)
 				.map(t2 -> {
 					OAuth2AuthorizeRequest.Builder builder = OAuth2AuthorizeRequest.withAuthorizedClient(authorizedClient).principal(t2.getT1());
-					if (t2.getT2().isPresent()) {
-						builder.attribute(ServerWebExchange.class.getName(), t2.getT2().get());
-					}
+					t2.getT2().ifPresent(exchange -> builder.attribute(ServerWebExchange.class.getName(), exchange));
 					return builder.build();
 				});
 	}
 
-	private Mono<Authentication> currentAuthentication() {
-		return ReactiveSecurityContextHolder.getContext()
-				.map(SecurityContext::getAuthentication)
-				.defaultIfEmpty(ANONYMOUS_USER_TOKEN);
-	}
-
-	private Mono<String> clientRegistrationId(Mono<Authentication> authentication) {
-		return authentication
-				.filter(t -> this.defaultOAuth2AuthorizedClient && t instanceof OAuth2AuthenticationToken)
-				.cast(OAuth2AuthenticationToken.class)
-				.map(OAuth2AuthenticationToken::getAuthorizedClientRegistrationId);
-	}
-
-	private Mono<ServerWebExchange> currentServerWebExchange() {
-		return Mono.subscriberContext()
-				.filter(c -> c.hasKey(ServerWebExchange.class))
-				.map(c -> c.get(ServerWebExchange.class));
-	}
-
 	private ClientRequest bearer(ClientRequest request, OAuth2AuthorizedClient authorizedClient) {
 		return ClientRequest.from(request)
 					.headers(headers -> headers.setBearerAuth(authorizedClient.getAccessToken().getTokenValue()))
 					.build();
 	}
 
+	/**
+	 * Sets the handler that handles authentication and authorization failures when communicating
+	 * to the OAuth2 Resource Server.
+	 *
+	 * <p>For example, a {@link RemoveAuthorizedClientReactiveOAuth2AuthorizationFailureHandler}
+	 * is typically used to remove the cached {@link OAuth2AuthorizedClient},
+	 * so that the same token is no longer used in future requests to the Resource Server.</p>
+	 *
+	 * <p>The failure handler used by default depends on which constructor was used
+	 * to construct this {@link ServerOAuth2AuthorizedClientExchangeFilterFunction}.
+	 * See the constructors for more details.</p>
+	 *
+	 * @param authorizationFailureHandler the handler that handles authentication and authorization failures.
+	 * @since 5.3
+	 */
+	public void setAuthorizationFailureHandler(ReactiveOAuth2AuthorizationFailureHandler authorizationFailureHandler) {
```

Comment:
IMHO, I don't think it makes sense to have an authorizationSuccessHandler here, and adding one might even cause confusion.

I say it might cause confusion, because someone might be tempted to set it to a `SaveAuthorizedClientReactiveOAuth2AuthorizationSuccessHandler`, which is not needed, and will just create contention within the backing authorized client repository/service.

I say it doesn't make sense, since it's almost impossible to tell if authorization was successful just by analyzing a response from a resource server.  For example, does an HTTP 400 error indicate that authorization was successful?  What about an HTTP 500 error?  Authorization might or might not have been successful when the resource server returns those status codes, depending on where in the resource server's stack the error occurred.

On the other side, it is immediately obvious when authorization fails, since the resource server will return 401 or 403.  So it makes sense to have a failure handler.

Reply:
Agreed, having a `setAuthorizationSuccessHandler()` might cause confusion here. We'll leave it out.